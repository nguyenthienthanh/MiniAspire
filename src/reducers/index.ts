import { combineReducers } from 'redux'
import loanReducers from './loan'
import loanRepaymentsReducer from './loanRepayments'
import loanRepaymentReducer from './loanRepayment'

const reducers = combineReducers({
  loan: loanReducers,
  loanRepayments: loanRepaymentsReducer,
  loanRepayment: loanRepaymentReducer,
})

export type ReduxReducers = Pick<
  ReturnType<typeof reducers>,
  'loan' | 'loanRepayments' | 'loanRepayment'
>

export default reducers
