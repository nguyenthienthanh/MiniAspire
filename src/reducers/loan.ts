import update from 'immutability-helper'
import {
  LoanReduxActionsType,
  FETCH_CURRENT_LOAN_REQUESTED,
  FETCH_CURRENT_LOAN_SUCCESS,
  FETCH_CURRENT_LOAN_FAILED,
  REPAY_LOAN_SUCCESS,
} from '../actions'
import { Loan } from '../types/loan'

export type LoanReduxState = {
  loading: boolean
  error?: Error
  data?: Loan
}

const initialState: LoanReduxState = {
  loading: false,
}

const loanReducer = (
  state: LoanReduxState = initialState,
  action: LoanReduxActionsType,
): LoanReduxState => {
  switch (action.type) {
    case REPAY_LOAN_SUCCESS:
      if (!state.data) {
        return state
      }

      return update(state, {
        data: {
          weeklyRepaid: {
            $set: true,
          },
        },
      })
    case FETCH_CURRENT_LOAN_REQUESTED:
      return update(state, {
        $set: {
          loading: true,
        },
      })
    case FETCH_CURRENT_LOAN_SUCCESS:
      return update(state, {
        $set: {
          loading: false,
          data: action.loan,
        },
      })
    case FETCH_CURRENT_LOAN_FAILED:
      return update(state, {
        $set: {
          loading: false,
          error: action.error,
        },
      })

    default:
      return state
  }
}

export default loanReducer
