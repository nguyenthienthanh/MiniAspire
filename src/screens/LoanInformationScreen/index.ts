export { default } from './LoanInformationScreen'
export * from './LoanInformationScreen'
