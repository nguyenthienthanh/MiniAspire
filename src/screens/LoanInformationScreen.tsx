import React, { FC } from 'react'
import { View, Text, StyleSheet } from 'react-native'
import { useSelector, useDispatch } from 'react-redux'
import { ReduxReducers } from '../reducers'
import { isEqual } from 'lodash'
import { fetchCurrentLoan, fetchLoanRepayment, repayLoan } from '../actions'
import Container from '../components/Container'

export type LoanInformationScreenProps = {}

const LoanInformationScreen: FC<LoanInformationScreenProps> = (props) => {
  // const loan = useSelector((state) => state, isEqual) as any

  // console.log(loan)

  // const dispatch = useDispatch()

  // React.useEffect(() => {
  //   dispatch(fetchCurrentLoan())
  // }, [dispatch])

  // React.useEffect(() => {
  //   dispatch(fetchLoanRepayment(loan.loan.data?.id))
  // }, [loan.loan.data?.id])

  // React.useEffect(() => {
  //   dispatch(repayLoan(loan.loan.data?.id))
  // }, [loan.loan.data?.id])

  return (
    <Container>
      <Text>LoanInformationScreen</Text>
    </Container>
  )
}

export default LoanInformationScreen

const styles = StyleSheet.create({})
