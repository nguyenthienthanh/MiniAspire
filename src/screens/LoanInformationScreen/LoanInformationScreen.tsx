import React, { FC, useCallback } from 'react'
import { StyleSheet, View, ActivityIndicator, FlatList } from 'react-native'

import { useSelector, useDispatch } from 'react-redux'
import { ReduxReducers } from '../../reducers'
import { isEqual } from 'lodash'
import { fetchCurrentLoan, fetchLoanRepayment } from '../../actions'
import Container from '../../components/Container'
import { Card, Text, colors, ListItem } from 'react-native-elements'
import { format } from 'date-fns'
import numeral from 'numeral'

export type LoanInformationScreenProps = {}

const LoanInformationScreen: FC<LoanInformationScreenProps> = (props) => {
  const loan = useSelector<ReduxReducers, ReduxReducers['loan']>(
    (state) => state.loan,
    isEqual,
  )

  const dispatch = useDispatch()

  React.useEffect(() => {
    dispatch(fetchCurrentLoan())
  }, [dispatch])

  return (
    <Container>
      <Card>
        <Card.Title testID="loanInformationTitle">Loan</Card.Title>
        <Card.Divider />

        {loan.loading && !loan.data ? (
          <ActivityIndicator />
        ) : (
          <>
            <Text testID="loanAmount" style={styles.text}>
              <Text style={styles.textTitle}>Amount: </Text>
              {numeral(loan.data?.amount ?? 0).format('0,0')} VND
            </Text>

            <Text testID="loanTerm" style={styles.text}>
              <Text style={styles.textTitle}>Term: </Text>
              {loan.data?.term &&
                format(new Date(loan.data.term), 'dd/MM/yyyy')}
            </Text>

            <Text testID="weeklyAmount" style={styles.text}>
              <Text style={styles.textTitle}>Weekly amount: </Text>
              {numeral(loan.data?.weeklyAmount ?? 0).format('0,0')} VND
            </Text>

            {loan.data?.weeklyRepaid && (
              <View testID="weeklyRepaid" style={styles.repaid}>
                <Text style={styles.repaidText}>REPAID</Text>
              </View>
            )}
          </>
        )}
      </Card>

      {loan.data?.id && <RecentRepayments loanId={loan.data.id} />}
    </Container>
  )
}

export default LoanInformationScreen

const RecentRepayments: FC<{ loanId: string }> = (props) => {
  const loanRepayments = useSelector<
    ReduxReducers,
    ReduxReducers['loanRepayments']
  >((state) => state.loanRepayments, isEqual)

  const dispatch = useDispatch()

  React.useEffect(() => {
    dispatch(fetchLoanRepayment(props.loanId))
  }, [dispatch, props.loanId])

  const keyExtractor = useCallback((_, index) => index.toString(), [])
  const renderItem = useCallback(({ item }) => {
    return (
      <ListItem testID="recentRepaymentItem">
        <ListItem.Content>
          <ListItem.Subtitle>
            <Text testID="recentRepaymentsAt" style={styles.text}>
              {format(new Date(item.repaymentsAt), 'dd/MM/yyyy')}
            </Text>
          </ListItem.Subtitle>

          <ListItem.Title>
            <Text testID="recentRepaymentAmount" style={styles.text}>
              <Text style={styles.textTitle}>Amount: </Text>
              {numeral(item.amount ?? 0).format('0,0')} VND
            </Text>
          </ListItem.Title>
        </ListItem.Content>
      </ListItem>
    )
  }, [])

  return (
    <Card>
      <Card.Title testID="recentRepaymentTitle">Recent repayments</Card.Title>
      <Card.Divider />
      {loanRepayments.loading && !loanRepayments.data.length ? (
        <ActivityIndicator />
      ) : (
        <FlatList
          data={loanRepayments.data}
          keyExtractor={keyExtractor}
          renderItem={renderItem}
        />
      )}
    </Card>
  )
}

const styles = StyleSheet.create({
  textTitle: {
    fontWeight: 'bold',
  },
  text: {
    marginBottom: 4,
  },
  repaid: {
    position: 'absolute',
    width: 80,
    height: 80,
    right: 0,
    bottom: 0,
    transform: [{ rotateZ: '30deg' }],
    justifyContent: 'center',
    alignContent: 'center',
    borderRadius: 40,
    borderWidth: 1,
    borderColor: colors.success,
  },
  repaidText: {
    fontSize: 18,
    textAlign: 'center',
    color: colors.success,
  },
})
