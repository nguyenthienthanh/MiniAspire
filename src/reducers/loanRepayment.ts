import update from 'immutability-helper'
import {
  LoanReduxActionsType,
  REPAY_LOAN_SUCCESS,
  REPAY_LOAN_REQUESTED,
  REPAY_LOAN_FAILED,
} from '../actions'
import { LoanRepayment } from '../types/loan'

export type LoanRepaymentReduxState = {
  loading: boolean
  error?: Error
  data?: LoanRepayment
}

const initialState: LoanRepaymentReduxState = {
  loading: false,
}

const loanRepaymentReducer = (
  state: LoanRepaymentReduxState = initialState,
  action: LoanReduxActionsType,
): LoanRepaymentReduxState => {
  switch (action.type) {
    case REPAY_LOAN_REQUESTED:
      return update(state, {
        loading: {
          $set: true,
        },
      })
    case REPAY_LOAN_SUCCESS:
      return update(state, {
        $set: {
          loading: false,
          data: action.loanRepayment,
        },
      })
    case REPAY_LOAN_FAILED:
      return update(state, {
        loading: {
          $set: false,
        },
        error: {
          $set: action.error,
        },
      })

    default:
      return state
  }
}

export default loanRepaymentReducer
