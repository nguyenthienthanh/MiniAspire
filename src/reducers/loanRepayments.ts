import update from 'immutability-helper'
import {
  FETCH_LOAN_REPAYMENTS_FAILED,
  FETCH_LOAN_REPAYMENTS_REQUESTED,
  FETCH_LOAN_REPAYMENTS_SUCCESS,
  LoanReduxActionsType,
  REPAY_LOAN_SUCCESS,
} from '../actions'
import { LoanRepayment } from '../types/loan'

export type LoanRepaymentReduxState = {
  loading: boolean
  error?: Error
  data: LoanRepayment[]
}

const initialState: LoanRepaymentReduxState = {
  loading: false,
  data: [],
}

const loanRepaymentsReducer = (
  state: LoanRepaymentReduxState = initialState,
  action: LoanReduxActionsType,
): LoanRepaymentReduxState => {
  switch (action.type) {
    case REPAY_LOAN_SUCCESS:
      return update(state, {
        data: { $unshift: [action.loanRepayment] },
      })
    case FETCH_LOAN_REPAYMENTS_REQUESTED:
      return update(state, {
        loading: {
          $set: true,
        },
      })
    case FETCH_LOAN_REPAYMENTS_SUCCESS:
      return update(state, {
        $set: {
          loading: false,
          data: action.loanRepayments,
        },
      })
    case FETCH_LOAN_REPAYMENTS_FAILED:
      return update(state, {
        $set: {
          loading: false,
          data: state.data,
          error: action.error,
        },
      })

    default:
      return state
  }
}

export default loanRepaymentsReducer
