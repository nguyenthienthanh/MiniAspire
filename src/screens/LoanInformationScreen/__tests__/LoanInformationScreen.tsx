/**
 * @format
 * @jest-environment jsdom
 */
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import React from 'react'
import 'react-native'
import * as Redux from 'react-redux'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import { ReduxReducers } from '../../../reducers'
import LoanInformationScreen from '../LoanInformationScreen'

Enzyme.configure({ adapter: new Adapter() })

const mockStore = configureStore()

describe('LoanInformationScreen', () => {
  it('should render correctly and match the snapshot', () => {
    expect.assertions(7)

    const loanRepayments: ReduxReducers['loanRepayments'] = {
      loading: true,
      data: [{ amount: 2000000, repaymentsAt: new Date('09/09/2020') }],
    }

    const wrapper = renderer.create(
      <Redux.Provider
        store={mockStore({
          loan: {
            loading: false,
            data: {
              id: 'loan-1',
              amount: 30000000,
              term: new Date('09/09/2020'),
              weeklyAmount: 2000000,
              weeklyRepaid: false,
            },
          },
          loanRepayments,
        })}
      >
        <LoanInformationScreen />
      </Redux.Provider>,
    )

    expect(
      wrapper.root.findByProps({ testID: 'loanInformationTitle' }).props
        .children,
    ).toBe('Loan')
    expect(
      wrapper.root.findByProps({ testID: 'loanAmount' }).props.children,
    ).toContain('30,000,000')
    expect(
      wrapper.root.findByProps({ testID: 'loanTerm' }).props.children,
    ).toContain('09/09/2020')
    expect(
      wrapper.root.findByProps({ testID: 'weeklyAmount' }).props.children,
    ).toContain('2,000,000')
    expect(
      wrapper.root.findByProps({ testID: 'recentRepaymentAmount' }).props
        .children,
    ).toContain('2,000,000')
    expect(
      wrapper.root.findByProps({ testID: 'recentRepaymentsAt' }).props.children,
    ).toContain('09/09/2020')
    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})
