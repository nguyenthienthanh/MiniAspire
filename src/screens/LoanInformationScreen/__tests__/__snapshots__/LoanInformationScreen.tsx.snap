// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoanInformationScreen should render correctly and match the snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgb(242, 242, 242)",
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "paddingTop": null,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#e1e8ee",
          "borderWidth": 1,
          "margin": 15,
          "marginBottom": 0,
          "padding": 15,
          "shadowColor": "rgba(0,0,0, .2)",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "transparent",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#43484d",
              "fontSize": 17.5,
              "fontWeight": "bold",
              "marginBottom": 15,
              "textAlign": "center",
            }
          }
          testID="loanInformationTitle"
        >
          Loan
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#bcbbc1",
              "height": 0.5,
              "marginBottom": 15,
            }
          }
        />
        <Text
          style={
            Object {
              "color": "#242424",
              "marginBottom": 4,
            }
          }
          testID="loanAmount"
        >
          <Text
            style={
              Object {
                "color": "#242424",
                "fontWeight": "bold",
              }
            }
          >
            Amount: 
          </Text>
          30,000,000
           VND
        </Text>
        <Text
          style={
            Object {
              "color": "#242424",
              "marginBottom": 4,
            }
          }
          testID="loanTerm"
        >
          <Text
            style={
              Object {
                "color": "#242424",
                "fontWeight": "bold",
              }
            }
          >
            Term: 
          </Text>
          09/09/2020
        </Text>
        <Text
          style={
            Object {
              "color": "#242424",
              "marginBottom": 4,
            }
          }
          testID="weeklyAmount"
        >
          <Text
            style={
              Object {
                "color": "#242424",
                "fontWeight": "bold",
              }
            }
          >
            Weekly amount: 
          </Text>
          2,000,000
           VND
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#e1e8ee",
          "borderWidth": 1,
          "margin": 15,
          "marginBottom": 0,
          "padding": 15,
          "shadowColor": "rgba(0,0,0, .2)",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "transparent",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#43484d",
              "fontSize": 17.5,
              "fontWeight": "bold",
              "marginBottom": 15,
              "textAlign": "center",
            }
          }
          testID="recentRepaymentTitle"
        >
          Recent repayments
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#bcbbc1",
              "height": 0.5,
              "marginBottom": 15,
            }
          }
        />
        <RCTScrollView
          data={
            Array [
              Object {
                "amount": 2000000,
                "repaymentsAt": 2020-09-08T17:00:00.000Z,
              },
            ]
          }
          disableVirtualization={false}
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={false}
          initialNumToRender={10}
          keyExtractor={[Function]}
          maxToRenderPerBatch={10}
          onContentSizeChange={[Function]}
          onEndReachedThreshold={2}
          onLayout={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={50}
          stickyHeaderIndices={Array []}
          updateCellsBatchingPeriod={50}
          viewabilityConfigCallbackPairs={Array []}
          windowSize={21}
        >
          <View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                testID="recentRepaymentItem"
              >
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#ffffff",
                      "borderColor": "#bcbbc1",
                      "flexDirection": "row",
                      "padding": 14,
                    }
                  }
                  testID="padView"
                >
                  <View
                    style={
                      Object {
                        "alignItems": "flex-start",
                        "flex": 1,
                        "justifyContent": "center",
                      }
                    }
                    theme={
                      Object {
                        "colors": Object {
                          "black": "#242424",
                          "disabled": "hsl(208, 8%, 90%)",
                          "divider": "#bcbbc1",
                          "error": "#ff190c",
                          "grey0": "#393e42",
                          "grey1": "#43484d",
                          "grey2": "#5e6977",
                          "grey3": "#86939e",
                          "grey4": "#bdc6cf",
                          "grey5": "#e1e8ee",
                          "greyOutline": "#bbb",
                          "platform": Object {
                            "android": Object {
                              "error": "#f44336",
                              "grey": "rgba(0, 0, 0, 0.54)",
                              "primary": "#2196f3",
                              "secondary": "#9C27B0",
                              "success": "#4caf50",
                              "warning": "#ffeb3b",
                            },
                            "ios": Object {
                              "error": "#ff3b30",
                              "grey": "#7d7d7d",
                              "primary": "#007aff",
                              "searchBg": "#dcdce1",
                              "secondary": "#5856d6",
                              "success": "#4cd964",
                              "warning": "#ffcc00",
                            },
                          },
                          "primary": "#2089dc",
                          "searchBg": "#303337",
                          "secondary": "#ca71eb",
                          "success": "#52c41a",
                          "warning": "#faad14",
                          "white": "#ffffff",
                        },
                      }
                    }
                  >
                    <Text
                      style={
                        Object {
                          "backgroundColor": "transparent",
                          "color": "#242424",
                          "fontSize": 15,
                        }
                      }
                      testID="listItemTitle"
                    >
                      <Text
                        style={
                          Object {
                            "color": "#242424",
                            "marginBottom": 4,
                          }
                        }
                        testID="recentRepaymentsAt"
                      >
                        09/09/2020
                      </Text>
                    </Text>
                    <Text
                      style={
                        Object {
                          "backgroundColor": "transparent",
                          "color": "#242424",
                          "fontSize": 17,
                        }
                      }
                      testID="listItemTitle"
                    >
                      <Text
                        style={
                          Object {
                            "color": "#242424",
                            "marginBottom": 4,
                          }
                        }
                        testID="recentRepaymentAmount"
                      >
                        <Text
                          style={
                            Object {
                              "color": "#242424",
                              "fontWeight": "bold",
                            }
                          }
                        >
                          Amount: 
                        </Text>
                        2,000,000
                         VND
                      </Text>
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
